### PR TITLE
DF 20736- GMx Glv EA

### PR DESCRIPTION
**Pull request by Subarna-Singh**
## Description
(https://smartcontract-it.atlassian.net/browse/DF-20736)
GMX Glv Token EA

## Changes
Added new composite adapter to fetch Glv contract prices

## Steps to Test
yarn test packages/composite/glv-token/test/

## Quality Assurance

- [ ] If a new adapter was made, or an existing one was modified so that its environment variables have changed, update the relevant `infra-k8s` configuration file.
- [x] If a new adapter was made, or an existing one was modified so that its environment variables have changed, update the relevant `adapter-secrets` configuration file or update the [soak testing blacklist](/packages/scripts/src/get-changed-adapters/soakTestBlacklist.ts).
- [x] If a new adapter was made, or a new endpoint was added, update the `test-payload.json` file with relevant requests.
- [ ] The branch naming follows git flow (`feature/x`, `chore/x`, `release/x`, `hotfix/x`, `fix/x`) or is created from Jira.
- [x] This is related to a maximum of one Jira story or GitHub issue.
- [x] Types are safe (avoid TypeScript/TSLint features like any and disable, instead use more specific types).
- [x] All code changes have 100% unit and integration test coverage. If testing is not applicable or too difficult to justify doing, the reasoning should be documented explicitly in the PR.

### CODE REVIEW

File: packages/composites/glv-token/package.json [hunk 1]
```diff
@@ -0,0 +1,42 @@
+{
+  "name": "@chainlink/glv-token-adapter",
+  "version": "1.1.11",
```
Comment (karen-stepanyan) [inline packages/composites/glv-token/package.json:3]:
```suggestion
  "version": "0.0.0",
```

File: packages/composites/glv-token/src/transport/price.ts [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+      const assetPromises = assets.map(async (asset) => {
+        const base = this.unwrapAsset(asset)
+        const requestConfig = {
+          url: source.url,
+          method: 'POST',
+          data: {
+            data: {
+              endpoint: 'crypto-lwba',
+              base,
+              quote: 'USD',
```
Comment (mxiao-cll) [inline packages/composites/glv-token/src/transport/price.ts:248]: Should this be a param?

File: packages/composites/glv-token/src/transport/price.ts [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+      })
+
+      promises.push(...assetPromises)
+    }
+
+    await Promise.all(promises)
+
+    this.validateRequiredResponses(priceProviders, sources, assets, dataRequestedTimestamp)
+
+    const medianValues = this.calculateMedian(assets, priceData)
```
Comment (mxiao-cll) [inline packages/composites/glv-token/src/transport/price.ts:285]: Can we re-use the Medianizer Composite Adapter?

File: packages/composites/glv-token/src/transport/utils.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+  return values.length % 2 ? values[half] : (values[half - 1] + values[half]) / 2
+}
+
+/*
+Formats a given number with a specified precision without leading zeros and decimal point.
+Price decimals = SIGNED_PRICE_DECIMALS - token decimals
+
+toFixed(14.84329267, 18) -> '14843292670000'
+toFixed(0.99999558, 6) -> '999995580000000000000000'
```
Comment (mxiao-cll) [inline packages/composites/glv-token/src/transport/utils.ts:27]: I don't really understand these two examples, should 18 and 6 be swapped?

File: packages/composites/glv-token/src/transport/utils.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  marketToken: string
+  indexToken: string
+  longToken: string
+  shortToken: string
+  isListed: boolean
+}
+
+export const glvMarkets = {
+  '0x528A5bac7E746C9A509A1f4F6dF58A03d44279F9': 'ETH/USDC',
+  '0xdF03EEd325b82bC1d4Db8b49c30ecc9E05104b96': 'WBTC/USDC',
```
Comment (mxiao-cll) [inline packages/composites/glv-token/src/transport/utils.ts:54]: Are these the only two supported input? We can put them in the the options of the input

File: packages/tsconfig.json [hunk 6]
```diff
@@ -44,6 +44,9 @@
     {
       "path": "./composites/dydx-rewards"
     },
+    {
+      "path": "./composites/gmlvtoken"
```
Comment (mxiao-cll) [inline packages/tsconfig.json:48]: Wrong name?

File: packages/composites/glv-token/src/transport/price.ts [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+    this.requester = dependencies.requester
+
+    this.glvReaderContract = new ethers.Contract(
+      adapterSettings.GLV_READER_CONTRACT_ADDRESS,
+      glvAbi,
+      this.provider,
+    )
+
+    await this.tokenInfo()
+    await this.marketInfo()
```
[the remark this answers is not in the document]
Comment (karen-stepanyan) [reply packages/composites/glv-token/src/transport/price.ts:81]: I think we should periodically call those APIs (maybe every hour or two) and update the mappings. If the initial calls fail we should crash the EA.

File: packages/composites/glv-token/src/transport/price.ts (hunk 3 above)
Comment (karen-stepanyan) [reply packages/composites/glv-token/src/transport/price.ts:285]: I would actually avoid using one more composite adapter as it overcomplicates the setup. Also the Medianizer is a v2 adapter and we should avoid using v2 EAs if possible

File: packages/composites/glv-token/src/transport/utils.ts (hunk 5 above)
Comment (karen-stepanyan) [reply packages/composites/glv-token/src/transport/utils.ts:54]: I really think we don't need this mapping. Only required input param should be GLV address. the contract returns both the long and short tokens

File: packages/composites/glv-token/src/transport/price.ts (hunk 2 above)
Comment (Subarna-Singh) [reply packages/composites/glv-token/src/transport/price.ts:246]: I dont understand your question

File: packages/composites/glv-token/src/transport/price.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Subarna-Singh) [reply packages/composites/glv-token/src/transport/price.ts:81]: Are we expecting such frequent updates in underlying contract values?

File: packages/composites/glv-token/src/transport/utils.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  const half = Math.floor(values.length / 2)
+
+  return values.length % 2 ? values[half] : (values[half - 1] + values[half]) / 2
+}
+
+/*
+Formats a given number with a specified precision without leading zeros and decimal point.
+Price decimals = SIGNED_PRICE_DECIMALS - token decimals
+
+toFixed(14.84329267, 6) -> '14843292670000'
```
Comment (karen-stepanyan) [inline packages/composites/glv-token/src/transport/utils.ts:26]: `toFixed(14.84329267, 6)` will result to 14843292670000000000000000. You need to swap arguments for 6 and 18

File: packages/composites/glv-token/src/transport/price.ts (hunk 2 above)
Comment (mxiao-cll) [reply packages/composites/glv-token/src/transport/price.ts:246]: I mean should we hard code USD here or pass it in via param instead

File: packages/composites/glv-token/src/transport/price.ts [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+              endpoint: 'crypto-lwba',
+              base,
+              quote: 'USD',
+            },
+          },
+        }
+
+        if (unsupportedAssets[source.name].includes(asset)) {
+          // skip assets not supported by EA to avoid Rate Limiting Error
+          undefined
```
Comment (karen-stepanyan) [inline packages/composites/glv-token/src/transport/price.ts:253]: ??

Comment (karen-stepanyan) [reply packages/composites/glv-token/src/transport/price.ts:253]: I don't really like this idea. We should always try to get the price from all sources and  `MIN_REQUIRED_SOURCE_SUCCESS` is just for some rare cases where DP is unavailable. Deliberately skipping some assets will cause issues sooner or later.

File: packages/composites/glv-token/src/transport/price.ts [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+    endpointName: string,
+    transportName: string,
+  ): Promise<void> {
+    await super.initialize(dependencies, adapterSettings, endpointName, transportName)
+    this.settings = adapterSettings
+    this.provider = new ethers.providers.JsonRpcProvider(
+      adapterSettings.ARBITRUM_RPC_URL,
+      adapterSettings.ARBITRUM_CHAIN_ID,
+    )
+    this.readerContract = new ethers.Contract(
```
Comment (mmcallister-cll) [inline packages/composites/glv-token/src/transport/price.ts:67]: is this used?

File: packages/composites/glv-token/src/transport/price.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (mxiao-cll) [reply packages/composites/glv-token/src/transport/price.ts:81]: I guess not frequent updates but right now the only way to update seems to be restart EA which is manual

File: packages/composites/glv-token/src/config/index.ts [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+      required: true,
+      default: 42161,
+    },
+    DATASTORE_CONTRACT_ADDRESS: {
+      description: 'Address of Data Store contract',
+      type: 'string',
+      required: true,
+      default: '0xFD70de6b91282D8017aA4E741e9Ae325CAb992d8',
+    },
+    READER_CONTRACT_ADDRESS: {
```
Comment (mmcallister-cll) [inline packages/composites/glv-token/src/config/index.ts:23]: is this used?

File: packages/composites/glv-token/src/endpoint/price.ts [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+  [
+    {
+      glv: '0x528A5bac7E746C9A509A1f4F6dF58A03d44279F9',
+    },
+  ],
+)
+
+export type BaseEndpointTypes = {
+  Parameters: typeof inputParameters.definition
+  Response: {
```
Comment (mmcallister-cll) [inline packages/composites/glv-token/src/endpoint/price.ts:23]: Could probably also do
```js
Response: SingleNumberResultResponse & {
  Data: {
    sources: Record<string, string[]>
  }
}
```

File: packages/composites/glv-token/src/transport/price.ts [hunk 13]
```diff
@@ -0,0 +1,380 @@
...
+      this.provider,
+    )
+
+    await this.tokenInfo()
+    await this.marketInfo()
+
+    // setInterval(() => {
+    //   this.tokenInfo()
+    //   this.marketInfo()
+    // }, 60 * 60 * 3 * 1000) // 3 hours
```
Comment (mmcallister-cll) [inline packages/composites/glv-token/src/transport/price.ts:86]: can this comment block be removed?

File: packages/composites/glv-token/src/config/readerAbi.json
Comment (mmcallister-cll) [inline packages/composites/glv-token/src/config/readerAbi.json:1]: READER_CONTRACT_ADDRESS doesn't seem to be used, can this be removed?

File: packages/composites/glv-token/src/transport/price.ts [hunk 14]
```diff
@@ -0,0 +1,380 @@
...
+  }
+
+  async tokenInfo() {
+    const requestConfig = {
+      url: this.settings.TOKEN_INFO_API,
+      method: 'GET',
+      data: {},
```
Comment (mmcallister-cll) [inline packages/composites/glv-token/src/transport/price.ts:93]: can remove if there is no data to pass

File: packages/composites/glv-token/src/index.ts [hunk 15]
```diff
@@ -0,0 +1,13 @@
+import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
+import { Adapter } from '@chainlink/external-adapter-framework/adapter'
+import { config } from './config'
+import { price } from './endpoint'
+
+export const adapter = new Adapter({
+  defaultEndpoint: price.name,
+  name: 'GLV',
```
Comment (mmcallister-cll) [inline packages/composites/glv-token/src/index.ts:8]: let's make this `GLV_TOKEN` to follow the standard convention

File: packages/composites/glv-token/src/transport/price.ts (hunk 13 above)
Comment (Subarna-Singh) [reply packages/composites/glv-token/src/transport/price.ts:86]: we need this. but its is interfering with the integration test. 
Waiting for Karen to get back on this issue.

**Review by karen-stepanyan — APPROVED** [review]

**Pull request merged**